File: honorsContract/h.py
#!/usr/bin/env python

from random import randint, shuffle, sample

INF = float('inf')

# input: a list
# output:   1 if there are any duplicates in the list
#           0 if the list contain exactly one of each element from 1 to len(list)
#           -1 otherwise
def canMakeRun(L):
    #print "CanMakeRun:", L, ":",
    setL = set(L)
    if len(L) > len(setL):
        #print "Too many elements"
        return 1
    return (setL == set(range(1, len(L)+1))) - 1

# L should be a range out of order. This will tell us the least amount of moves
# to combine the dolls
def solveSingleSet(L, debug = False):
    if len(L) == 1:
        if debug: print L, ":", 0
        return 0
    minLeft = minRight = float('inf')
    maximum = L[0]
    maxIndex = 0
    for index, num in enumerate(L[1:]):
        if num > maximum:
            minLeft = min(minLeft, maximum, minRight)
            minRight = float('inf')
            maximum = num
            maxIndex = index+1
        elif num < minRight:
            minRight = num

    if minLeft == float('inf') or minRight == float('inf'):
        val = 1 + solveSingleSet(L[:maxIndex]+L[maxIndex+1:], debug)
        print L, ":", val
        return val
    else:
        minSide = max(minLeft, minRight)
        leftOpens = rightOpens = 0
        for num in L[:maxIndex]:
            if num >= minSide:
                leftOpens+=1
        for num in L[maxIndex+1:]:
            if num >= minSide:
                rightOpens+=1
        val = solveSingleSet(L[:maxIndex], debug) + solveSingleSet(L[maxIndex+1:], debug) + leftOpens + rightOpens + 2
        print L, ":", val
        return val
        # [(5, 10, 6, 17, 15, <1>, 4, 11, 16, 18,  7), 19,(12, 9,  3,  8, 14, <2>,  13)]
        # [(5, 10, 6, 17, 15, <1>, 4, 11, 16),18,(<7>)]  [(12, 9, <3>, 8,)14,(<2>,  13)]

def solve(L, debug = False):
    runs = [[i, i+1] for i in range(len(L))]
    return solveRecurse(L, runs, debug)

def solveRecurse(L, runs, debug = False):
    # group going right
    groupDirection(L, runs, 1, debug)
    # group going left
    choice = groupDirection(L, runs, -1, debug)

    if isinstance(choice, str):
        return choice
    if choice == -1:
        print "There are no choices"
        return sum(map(solveSingleSet, (getList(L, i, i) for i in runs), [True]*len(runs)))
    else:
        print "Choice at index", choice
        rightRuns = [run[:] for run in runs]
        val = rightRuns.pop(choice)
        rightRuns[choice][0] = val[0]
        right = solveRecurse(L, rightRuns, debug)
        print "Right:", right

        #if choice > 0 and not canMakeRun(runs[choice-1]):
        leftRuns = [run[:] for run in runs]
        val = leftRuns.pop(choice)
        leftRuns[choice-1][1] = val[1]
        left = solveRecurse(L, leftRuns, debug)
        print "Left:", left
        #else:
            #left = INF
            #print "Skipping Left"

        return min(left, right)

def groupDirection(L, runs, direction, debug = False):
    if direction == 1:
        start = 0
        end = lambda : len(runs)
    elif direction == -1:
        start = len(runs)-1
        end = lambda: -1
    else:
        raise Exception("Direction must be one of 1 and -1: " + str(direction))

    bothWays = -1

    print "Starting in the ", direction, " direction"
    printList(L, runs)
    index = start
    while index != end():
        #if debug: printList(L, runs)

        # the indexes we can continue to
        rightIndex = leftIndex = index

        # continue to the right until we reach the end or are not not finished yet
        while rightIndex < len(runs) and canMakeRun(getList(L, runs[index], runs[rightIndex])) < 0:
            rightIndex += 1
        # if we reached the end or have too many elements
        if rightIndex == len(runs) or canMakeRun(getList(L, runs[index], runs[rightIndex])) != 0:
            rightIndex = index

        # continue to the left until we reach the beginning or are not not finished yet
        while leftIndex >= 0 and canMakeRun(getList(L, runs[leftIndex], runs[index])) < 0:
            leftIndex -= 1
        # if we reached the beginning or have too many elements
        if leftIndex == -1 or canMakeRun(getList(L, runs[leftIndex], runs[index])) != 0:
            #print "Changed left index"
            leftIndex = index

        #if debug:
            #print "Current Run:"
            #printRun(L, *runs[index])
            #print "Right run:"
            #printRun(L, runs[index][0], runs[rightIndex][1])
            #print "Left run:"
            #printRun(L, runs[leftIndex][0], runs[index][1])

        # if we can continue right
        if rightIndex > index:
            # if we can continue left
            if leftIndex < index:
                # this can extend both ways
                bothWays = index
                index += direction
            else:
                # join right
                val = runs.pop(index)
                runs[index][0] = val[0]
                if bothWays > index: bothWays-=1
        else:
            # if we can continue left
            if leftIndex < index:
                # join left
                val = runs.pop(index)
                index -= 1
                runs[index][1] = val[1]
                if bothWays > index: bothWays-=1
            else:
                # if the current run is complete
                if canMakeRun(getList(L, runs[index], runs[index])) == 0:
                    index += direction
                # if the first run or last run cannot complete
                elif index == 0 or index == len(runs)-1:
                    return "Not Solvable"
                else:
                    canMakeRunVal = canMakeRun(getList(L, runs[index-1], runs[index+1]))
                    # if the extended run has too many elements
                    if canMakeRunVal == 1:
                        return "Not Solvable"
                    # if the extended run does not enough elements or is complete
                    else:
                        runs.pop(index)
                        val = runs.pop(index)
                        index -= 1
                        runs[index][1] = val[1]
                        if bothWays > index: bothWays -= 2
        #if debug:
            #printList(L, runs)
            #raw_input()
            #print
            #print

    printList(L, runs)
    print

    # if the list is solvable, return the last instance there is a choice
    return bothWays

def printList(L, runs):
    print L
    print '[' + '><'.join(' '*(1+3*(i[1]-i[0]-1)) for i in runs) + ']'

def printRun(L, left, right):
    print L
    print ' '*3*left + '<' + ' '*(3*(right-left)-2) + '>'

def getList(L, left, right):
    return L[left[0]:right[1]]

if __name__ == "__main__":
    debug = True

    L = []
    if debug:
        num =  randint(2, 10)
        print num, "sets"
        for i in range(num):
            A = list(range(1,randint(4, 6)))
            shuffle(A)
            L.extend(A)
    else:
        pass
        #L = [1, 2, 4, 3, 1, 2, 5, 6, 4, 3, 5, 4, 1, 3, 2, 4, 1, 3, 2]
        #L = [1, 2, 3, 1, 3, 2, 1, 3, 2, 1, 4, 2, 3]
        #L = [3, 2, 1, 3, 1, 2, 1, 3, 4, 2, 3, 1, 2]
        #L = [1, 3, 2, 2, 3, 4, 1, 6, 5, 3, 4, 6, 1, 5, 2, 3, 1, 4, 2]
        #L = [1, 4, 5, 2, 3, 6, 4, 1, 5, 3, 2]
        #L = [3, 1, 6, 5, 4, 2, 1, 2, 6, 3, 4, 5, 5, 1, 4, 2, 3]
        #L = [3, 2, 1, 1, 3, 2, 4, 2, 3, 5, 6, 1]

        #L = [2, 4, 3]
        #L = [1, 2, 1, 2, 4, 3, 3]
        #L = [1, 2, 4]

        #L = [3, 2, 1, 4, 5, 3, 2, 1]
        L = [3, 2, 1, 4, 5, 6, 7, 4, 3, 2, 1, 5, 4, 3, 2, 1]

    print "Dolls:", L
    print "Sorted:"
    print solve(L, debug)

    #print solveSingleSet(sample(range(1, 20), 19), debug)

'''
3 2 1 3 1 2 1 3 4 2 3 1 2
<   > <   > <     > <   >

1 3 2 2 3 4 1 6 5 3 4 6 1 5 2 3 1 4 2
<   > <         > <         > <     >

1 4 5 2 3  6   4 1 5 3 2
<       > {<>} <       >

3 1 6 5 4 2 1 2 6 3 4 5 5 1 4 6 2 3
<         > <         > <         >

3, 2, 1, 1, 3, 2, 4, 2, 3, 5, 6, 1
<     >  <     >  <              >
'''

File: honorsContract/h.cpp
#include <algorithm>
#include <iostream>
#include <list>
#include <set>
#include <list>
#include <utility>
#include <random>
using namespace std;

typedef list<pair<int, int> >::iterator iter;

void printArray(const int L[], const int size) {
    cout << "[" << L[0];
    for(int i=1; i<size; i++){
        cout << ", " << L[i];
    }
    cout << "]" << endl;
}
void printRuns(const list<pair<int, int> >& runs){
    cout << '[';
    for(list<pair<int, int> >::const_iterator index = runs.begin(); index != runs.end(); index++) {
        if(index != runs.begin()){
            cout << "><";
        }
        for(int i=0; i < 1+3*(index->second - index->first - 1); i++) {
            cout << ' ';
        }
    }
    cout << "]" << endl;
}
void printCurrentIndex(list<pair<int, int> >& runs, iter index){
    cout << ' ';
    for(list<pair<int, int> >::const_iterator temp_index = runs.begin(); temp_index != runs.end(); temp_index++) {
        if(temp_index == index){
            cout << "_";
            break;
        }
        for(int i=0; i < 3*(temp_index->second - temp_index->first); i++) {
            cout << ' ';
        }
    }
    cout << endl;
}
int canMakeRun(list<int> L);
list<int> getList(int L[], pair<int, int> left, pair<int, int> right);
void joinRun(list<pair<int, int> >& runs, iter& from, iter& to);
int solveSingleSet(int L[], const int begin, const int end);
int solve(int L[], const int size);
int solveRecursive(int L[], const int size, list<pair<int, int> >& runs, const int sizeRuns);
iter groupForward(int L[], const int size, list<pair<int, int> >& runs);
iter groupBackward(int L[], const int size, list<pair<int, int> >& runs);

const int INF = 2000000000;

// input: a list
// output:   1 if there are any duplicates in the list
//           0 if the list contain exactly one of each element from 1 to list.size()
//           -1 otherwise
int canMakeRun(list<int> L) {
    int size = L.size();
    L.sort();
    L.unique();

    if (L.size() < size) {
        // too many elements
        return 1;
    }
    int value = 1;
    for(auto i=L.begin(); i != L.end(); i++) {
        if (*i != value) {
            // one of the numbers is missing
            return -1;
        }
        value++;
    }
    // the list contains all the number from 1 to size
    return 0;
}

// gets a list from the first element of the first pair and the second element of the second pair
list<int> getList(int L[], pair<int, int> left, pair<int, int> right) {
    return list<int>(L+left.first, L+right.second);
}

void joinRun(list<pair<int, int> >& runs, iter& from, iter& to) {
    int first = from->first;
    runs.erase(from, to);
    to->first = first;
    from = to;
}

int solveSingleSet(int L[], const int begin, const int end) {
    if(end - begin <= 1) {
        return 0;
    }
    int minLeft = INF;
    int minRight = INF;
    int maximum = L[begin];
    int maxIndex = begin;
    for(int i=begin+1; i<end; i++) {
        if(L[i] > maximum) {
            minLeft = min(min(minLeft, maximum), minRight);
            minRight = INF;
            maximum = L[i];
            maxIndex = i;
        }
        else if(L[i] < minRight) {
            minRight = L[i];
        }
    }
    if(minLeft == INF) {
        return 1 + solveSingleSet(L, begin+1, end);
    }
    else if(minRight == INF) {
        return 1 + solveSingleSet(L, begin, end-1);
    }
    else {
        int minSide = max(minLeft, minRight);
        int leftOpens = 0;
        int rightOpens = 0;
        for(int i=begin; i<maxIndex; i++) {
            if(L[i] >= minSide){
                leftOpens++;
            }
        }
        for(int i=maxIndex+1; i<end; i++) {
            if(L[i] >= minSide){
                rightOpens++;
            }
        }
        return solveSingleSet(L, begin, maxIndex) + solveSingleSet(L, maxIndex+1, end) + leftOpens + rightOpens + 2;
    }
}

int solve(int L[], const int size) {
    // create the list of runs
    list<pair<int, int> > runs;
    for (int i=0; i<size; i++) {
        runs.push_back(make_pair(i, i+1));
    }

    return solveRecursive(L, size, runs, size);
}

int solveRecursive(int L[], const int size, list<pair<int, int> >& runs, const int sizeRuns) {
    printArray(L, size);
    printRuns(runs);
    cout << endl;
    groupForward(L, size, runs);
    auto choice = groupBackward(L, size, runs);

    if(choice == runs.end()) {
        cout << "There are no choices" << endl;
        int sum = 0;
        for(auto i = runs.begin(); i != runs.end(); i++) {
            int val = solveSingleSet(L, i->first, i->second);
            printArray(L, size);
            printRuns(runs);
            cout << "   " << val << endl;
            sum += val;
        }
        return sum;
    }
    else {
        cout << "There is a choice" << endl;
        auto rightRuns = runs;
        auto rightChoice = rightRuns.begin();
        while(*rightChoice != *choice) {
            rightChoice++;
        }
        auto right = rightChoice;
        right++;
        joinRun(rightRuns, rightChoice, right);
        printArray(L, size);
        printRuns(rightRuns);
        int rightVal = solveRecursive(L, size, rightRuns, sizeRuns-1);

        auto leftRuns = runs;
        auto leftChoice = leftRuns.begin();
        while(*leftChoice != *choice) {
            leftChoice++;
        }
        auto left = leftChoice;
        left--;
        joinRun(leftRuns, left, leftChoice);
        printArray(L, size);
        printRuns(leftRuns);
        int leftVal = solveRecursive(L, size, leftRuns, sizeRuns-1);

        return min(rightVal, leftVal);
    }
}

iter groupForward(int L[], const int size, list<pair<int, int> >& runs) {
    // start is the element before the first element
    auto beforeBegin = --runs.begin();
    auto index = runs.begin();
    auto beforeEnd = --runs.end();
    auto choice = runs.end();

    while(index != runs.end()) {
        printArray(L, size);
        printRuns(runs);
        printCurrentIndex(runs, index);

        // the indexes we can continue to
        auto rightIndex = index;
        auto leftIndex = index;

        // continue to the right until we reach the end or are not not finished yet
        while (rightIndex != runs.end() and canMakeRun(getList(L, *index, *rightIndex)) < 0) {
            rightIndex++;
        }
        // if we reached the end or have too many elements
        if (rightIndex == runs.end() or canMakeRun(getList(L, *index, *rightIndex)) != 0) {
            rightIndex = index;
        }

        // continue to the left until we reach the beginning or are not not finished yet
        while (leftIndex != beforeBegin and canMakeRun(getList(L, *leftIndex, *index)) < 0) {
            leftIndex--;
        }
        // if we reached the beginning or have too many elements
        if (leftIndex == beforeBegin or canMakeRun(getList(L, *leftIndex, *index)) != 0) {
            leftIndex = index;
        }

        // if we can continue right
        if (rightIndex != index) {
            // if we can continue left
            if (leftIndex != index) {
                // this can extend both ways
                //bothWays = index;
                choice = index;
                index++;
            }
            else {
                // join right
                auto old_index = index;
                index++;
                joinRun(runs, old_index, index);
            }
        }
        else {
            // if we can continue left
            if (leftIndex != index) {
                // join left
                auto old_index = index;
                index--;
                joinRun(runs, index, old_index);
            }
            else {
                // if the current run is complete
                if (canMakeRun(getList(L, *index, *index)) == 0) {
                    index++;
                }
                // if the first run or last run cannot complete
                else if (index == runs.begin() or index == beforeEnd) {
                    throw "Not Solvable";
               }
                else {
                    auto right_of_index = index;
                    right_of_index++;
                    index--;
                    int canMakeRunVal = canMakeRun(getList(L, *index, *right_of_index));
                    // if the extended run has too many elements
                    if (canMakeRunVal == 1) {
                        throw "Not Solvable";
                    }
                    // if the extended run does not enough elements or is complete
                    else {
                        joinRun(runs, index, right_of_index);
                    }
                }
            }
        }
        printRuns(runs);
        cout << endl;
    }
    return choice;
}

iter groupBackward(int L[], const int size, list<pair<int, int> >& runs) {
    auto beforeBegin = --runs.begin();
    auto index = --runs.end();
    auto beforeEnd = --runs.end();
    auto choice = --runs.begin();

    while(index != beforeBegin) {
        printArray(L, size);
        printRuns(runs);
        printCurrentIndex(runs, index);

        // the indexes we can continue to
        auto rightIndex = index;
        auto leftIndex = index;

        // continue to the right until we reach the end or are not not finished yet
        while (rightIndex != runs.end() and canMakeRun(getList(L, *index, *rightIndex)) < 0) {
            rightIndex++;
        }
        // if we reached the end or have too many elements
        if (rightIndex == runs.end() or canMakeRun(getList(L, *index, *rightIndex)) != 0) {
            rightIndex = index;
        }

        // continue to the left until we reach the beginning or are not not finished yet
        while (leftIndex != beforeBegin and canMakeRun(getList(L, *leftIndex, *index)) < 0) {
            leftIndex--;
        }
        // if we reached the beginning or have too many elements
        if (leftIndex == beforeBegin or canMakeRun(getList(L, *leftIndex, *index)) != 0) {
            leftIndex = index;
        }

        // if we can continue right
        if (rightIndex != index) {
            // if we can continue left
            if (leftIndex != index) {
                // this can extend both ways
                choice = index;
                index--;
            }
            else {
                // join right
                auto old_index = index;
                index++;
                joinRun(runs, old_index, index);
            }
        }
        else {
            // if we can continue left
            if (leftIndex != index) {
                // join left
                auto old_index = index;
                index--;
                joinRun(runs, index, old_index);
            }
            else {
                // if the current run is complete
                if (canMakeRun(getList(L, *index, *index)) == 0) {
                    index--;
                }
                // if the first run or last run cannot complete
                else if (index == runs.begin() or index == beforeEnd) {
                    throw "Not Solvable";
                }
                else {
                    auto left_of_index = index;
                    left_of_index--;
                    index++;
                    int canMakeRunVal = canMakeRun(getList(L, *left_of_index, *index));
                    // if the extended run has too many elements
                    if (canMakeRunVal == 1) {
                        throw "Not Solvable";
                    }
                    // if the extended run does not enough elements
                    else {
                        joinRun(runs, left_of_index, index);
                    }
                }
            }
        }
        printRuns(runs);
        cout << endl;
    }
    return choice;
}

int main() {

    //int size;
    //cin >> size;
    //int* L = new int[size];
    //for(int i=0; i < size; i++) {
        //cin >> L[i];
    //}
    
    srand(time(NULL));
    for(int count=0; count<100; count++) {
        int size = rand()%16 + 15;
        int* L = new int[size];

        int index = 0;
        while(index < size) {
            int new_size = min(rand()%3 + 3, size - index);
            for(int i=0; i < new_size; i++) {
                L[index+i] = i+1;
            }
            random_shuffle(L+index, L+index+new_size);
            index += new_size;
        }

        cout << "Dolls: ";
        printArray(L, size);

        try {
            cout << solve(L, size) << endl;
        }
        catch (const char* error) {
            cout << "impossible" << endl;
        }

        delete [] L;
        
        cin.get();
    }

    return 0;
}
